### PR TITLE
[x/programs] Fix simulator rewrite

### PR DESCRIPTION
**Pull request by iFrostizz**
Followup on https://github.com/ava-labs/hypersdk/pull/900
- The serialization trick that is happening in the `_guest` functions expanded by the `public` attribute macro is not aligned with the `params!` macro. This is why I manually serialized the params in the counter example. Let me know if it should rather be an ignored test.
- I'm not sure if we should write the data to a new scratch space when deleting stuff from the storage, you can see in the `delete` link that it's returning the data that is currently sitting in memory (but is passed as a pointer still), could this data go out of scope ?

### CODE REVIEW

File: x/programs/cmd/simulator/build.rs [hunk 1]
```diff
@@ -24,7 +24,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
 
     let current_dir = std::env::current_dir().unwrap();
 
-    let simulator_path = "bin/simulator";
+    let simulator_path = "../../v2/bin/simulator";
```
Comment (iFrostizz) [inline x/programs/cmd/simulator/build.rs:27]: Are we leaving this as-is?

File: x/programs/v2/cmd/plan.go [hunk 2]
```diff
@@ -192,7 +192,7 @@ func (c *runCmd) RunStep(ctx context.Context, db *state.SimpleMutable) (*Respons
 	resp := newResponse(index)
 	err = runStepFunc(ctx, c.log, db, step.Endpoint, step.MaxUnits, step.Method, params, step.Require, resp)
 	if err != nil {
-		c.log.Debug("simulation step", zap.Error(err))
+		c.log.Error("simulation step err", zap.Error(err))
```
Comment (iFrostizz) [inline x/programs/v2/cmd/plan.go:195]: I'm wondering if the simulation should halt if there is an error, both cases kinda make sense

Comment (dboehm-avalabs) [reply x/programs/v2/cmd/plan.go:195]: I think there may be cases where we want the error to be returned without killing the whole simulator.

Comment (iFrostizz) [reply x/programs/v2/cmd/plan.go:195]: True, this would be part of https://github.com/ava-labs/hypersdk/issues/595 (I think)

File: scripts/tests.unit.sh [hunk 3]
```diff
@@ -17,4 +17,4 @@ source "$HYPERSDK_PATH"/scripts/constants.sh
 
 # Provision of the list of tests requires word splitting, so disable the shellcheck
 # shellcheck disable=SC2046
-go test -race -timeout="3m" -coverprofile="coverage.out" -covermode="atomic" $(find . -name "*.go" | grep -v "./x/programs/cmd" | grep -v "./examples/morpheusvm" | grep -v "./examples/tokenvm" | xargs -n1 dirname | sort -u | xargs)
+go test -race -timeout="3m" -coverprofile="coverage.out" -covermode="atomic" $(find . -name "*.go" | grep -v "./x/programs/cmd" | grep -v "./x/programs/v2/cmd" | grep -v "./examples/morpheusvm" | grep -v "./examples/tokenvm" | xargs -n1 dirname | sort -u | xargs)
```
Comment (richardpringle) [inline scripts/tests.unit.sh:20]: Do we not want to run the `cmd` tests? If there are test files, we should probably run them, right? If we want to trigger all the tests from Rust, we can just not write any test-files.

File: x/programs/cmd/simulator/build.rs [hunk 4]
```diff
@@ -9,7 +9,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
     let go_mod_download_output = Command::new("go").args(["mod", "download"]).output()?;
 
     if !go_mod_download_output.status.success() {
-        println!("cargo:warning=go mod downlod stdout:");
+        println!("cargo:warning=go mod download stdout:");
```
Comment (richardpringle) [inline x/programs/cmd/simulator/build.rs:12]: 😅  who did that

File: x/programs/cmd/simulator/build.rs (hunk 1 above)
Comment (richardpringle) [reply x/programs/cmd/simulator/build.rs:27]: For now. Step 1, get tests working again. Step 2, rename.

File: x/programs/examples/utils.go
Comment (richardpringle) [inline x/programs/examples/utils.go:1]: What are we still using from this file? I would prefer we didn't use anything in an `examples` directory. 

I guess this is a bit of a nit, so maybe we can do it in a follow up

File: x/programs/rust/examples/counter/src/lib.rs [hunk 5]
```diff
@@ -190,7 +188,7 @@ mod tests {
     }
 
     #[test]
-    #[ignore]
+    #[ignore = "need to fix params macro"]
```
Comment (richardpringle) [inline x/programs/rust/examples/counter/src/lib.rs:191]: I think we can probably get rid of the `params` macro and solve this in #723

File: x/programs/rust/examples/token/src/lib.rs
Comment (richardpringle) [inline x/programs/rust/examples/token/src/lib.rs:1]: 🙌

File: x/programs/rust/sdk-macros/src/lib.rs [hunk 6]
```diff
@@ -138,7 +138,7 @@ pub fn public(_: TokenStream, item: TokenStream) -> TokenStream {
             #[derive(borsh::BorshDeserialize)]
             struct Args {
                 #(#args_props),*
-             }
+            }
```
Comment (richardpringle) [inline x/programs/rust/sdk-macros/src/lib.rs:141]: 😞

File: x/programs/rust/wasmlanche-sdk/src/params.rs [hunk 7]
```diff
@@ -17,7 +17,8 @@ macro_rules! params {
 pub struct Param(Vec<u8>);
 
 /// A collection of [borsh] serialized parameters.
-pub struct Params(Vec<u8>);
+// TODO remove this pub attribute once the params! case is resolved
```
Comment (richardpringle) [inline x/programs/rust/wasmlanche-sdk/src/params.rs:20]: I think #723 makes this obsolete.

File: x/programs/v2/cmd/custom_imports/README.md
Comment (richardpringle) [inline x/programs/v2/cmd/custom_imports/README.md:1]: Please just delete this README, it's completely useless. If we want to do the TODO, we can create an issue.

File: x/programs/v2/cmd/errors.go
Comment (richardpringle) [inline x/programs/v2/cmd/errors.go:1]: This shouldn't be moved into the v2 directory

File: x/programs/v2/cmd/interpreter.go
Comment (richardpringle) [inline x/programs/v2/cmd/interpreter.go:1]: Same comment about moving the files.

File: x/programs/v2/cmd/plan.go [hunk 8]
```diff
@@ -80,15 +79,20 @@ func (c *runCmd) Happened() bool {
...
 func (c *runCmd) Init() (err error) {
 	var planStep []byte
-	if c.planStep != nil && len(*c.planStep) > 0 {
-		planStep = []byte(*c.planStep)
-	} else if len(*c.file) > 0 {
-		// read simulation step from file
-		planStep, err = os.ReadFile(*c.file)
-		if err != nil {
-			return err
+	switch {
```
Comment (richardpringle) [inline x/programs/v2/cmd/plan.go:82]: 🥇

File: x/programs/v2/cmd/plan.go (hunk 2 above)
Comment (richardpringle) [reply x/programs/v2/cmd/plan.go:199]: Agreed with both!

File: x/programs/v2/cmd/testdata/basic_token.yaml
Comment (richardpringle) [inline x/programs/v2/cmd/testdata/basic_token.yaml:1]: We should get rid of the yaml support entirely

Comment (richardpringle) [reply x/programs/v2/cmd/testdata/basic_token.yaml:1]: Or only support yaml since it's a superset of json anyway

Comment (richardpringle) [reply x/programs/v2/cmd/testdata/basic_token.yaml:1]: but I definitely recommend just getting rid of it, the original intent for the simulator was to run it in a more manual fashion and we aren't doing that anymore.

File: x/programs/v2/runtime/config.go [hunk 9]
```diff
@@ -1,7 +1,7 @@
 // Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
 // See the file LICENSE for licensing terms.
 
-package v2
+package runtime
```
Comment (richardpringle) [inline x/programs/v2/runtime/config.go:4]: 🤔  I wouldn't have expected the rename to happen in your PR... Wouldn't it just be better to get rid of the whole v2 package/folder in one shot where that's all that's happening in the PR?

File: x/programs/v2/vm/actions/program_execute.go [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+	// Returning -1, -1 means that the action is always valid.
+	return -1, -1
+}
+
+// CallParam defines a value to be passed to a guest function.
+type CallParam struct {
+	Value interface{} `json,yaml:"value"`
+}
+
+func SerializeParams(p []CallParam) ([]byte, error) {
```
Comment (richardpringle) [inline x/programs/v2/vm/actions/program_execute.go:143]: @dboehm-avalabs, we've talked about this, but I don't remember what the conclusion was. *I think* we said that it would make sense for the user to serialize the params before constructing the action, right?

File: x/programs/v2/vm/actions/program_execute.go
Comment (richardpringle) [inline x/programs/v2/vm/actions/program_execute.go:1]: Don't the actions depend on the VM? It seems like this dependency is reversed... It should at least be a peer to the VM, I think. 

Let me know if you need me to explain further.

**Review by richardpringle — CHANGES_REQUESTED** [review]
Sorry, part of my change-requests are on me. *I think* there could be a cleaner separation between the simulator logic and the engine logic. The engine (currently the `v2` package) should be a dependency of the `simulator`. 

While this may seem like `nit` territory, I really want to avoid the mess we had before where the line was not clear. 

Interested in hearing @dboehm-avalabs's thoughts on the subject as well.

File: x/programs/v2/vm/actions/program_execute.go (hunk 10 above)
Comment (dboehm-avalabs) [reply x/programs/v2/vm/actions/program_execute.go:143]: I agree, though we do need some preprocessing of the params somewhere.  We currently support address aliasing in the simulator steps, so we need to process the params as a list *before* they get serialized.

File: x/programs/v2/runtime/config.go (hunk 9 above)
Comment (dboehm-avalabs) [reply x/programs/v2/runtime/config.go:4]: Yeah, I'll probably do that in a separate PR after https://github.com/ava-labs/hypersdk/pull/917

File: x/programs/cmd/simulator/build.rs (hunk 4 above)
Comment (iFrostizz) [reply x/programs/cmd/simulator/build.rs:12]: I have no idea, my `git blame` is broken!!

File: x/programs/v2/cmd/custom_imports/README.md
Comment (iFrostizz) [reply x/programs/v2/cmd/custom_imports/README.md:1]: deleted!

File: x/programs/v2/cmd/testdata/basic_token.yaml
Comment (iFrostizz) [reply x/programs/v2/cmd/testdata/basic_token.yaml:1]: Nuking yaml support, it makes sense and also improves code performance.

File: x/programs/v2/runtime/config.go (hunk 9 above)
Comment (iFrostizz) [reply x/programs/v2/runtime/config.go]: @dboehm-avalabs I think that you solved this by moving stuff from `v2` to `runtime` ?

File: x/programs/v2/vm/actions/program_execute.go
Comment (iFrostizz) [reply x/programs/v2/vm/actions/program_execute.go:1]: We will remove any idea of actions in a followup PR, they don't make sense in the context of the simulator as we discussed.

File: x/programs/cmd/simulator/vm/actions/program_execute.go [hunk 11]
```diff
@@ -60,80 +66,35 @@ func (t *ProgramExecute) Execute(
 	_ int64,
 	actor codec.Address,
 	_ ids.ID,
-) (success bool, computeUnits uint64, output []byte, err error) {
```
Comment (iFrostizz) [inline x/programs/cmd/simulator/vm/actions/program_execute.go:63]: The reason of this is https://github.com/ava-labs/hypersdk/pull/858/commits/b67c5f07936fc758fc78b90f702449e3e7f29185

**Review by dboehm-avalabs — APPROVED** [review]

File: consts/consts.go [hunk 12]
```diff
@@ -7,6 +7,7 @@ const (
 	// These `codec` consts are defined here to avoid a circular dependency
 	BoolLen   = 1
 	ByteLen   = 1
+	IDLen     = 32
```
Comment (patrick-ogrady) [inline consts/consts.go:10]: Don't add this back in.

Comment (patrick-ogrady) [reply consts/consts.go:10]: use `ids.IDLen`

Comment (iFrostizz) [reply consts/consts.go]: Thanks! Removed

File: x/programs/runtime/import_log_release.go
Comment (iFrostizz) [inline x/programs/runtime/import_log_release.go:1]: The reason for this is that build flags are currently broken, we cannot override the default one with `debug` for instance, and will be fixed later.

**Review by richardpringle — CHANGES_REQUESTED** [review]
Can you make sure you reply to all my comments? The only thing that I actually think should change is the `CallProgramArgs`. I think you should put that code back, especially since you're ignoring that test.

And I see that you pushed a change just now... 😞

File: x/programs/runtime/import_state.go [hunk 13]
```diff
@@ -74,7 +74,26 @@ func NewStateAccessModule() *ImportModule {
...
+				}
+
+				bytes = append([]byte{1}, bytes...)
+
+				err = callInfo.State.Remove(ctx, key)
+				if err != nil {
+					return nil, err
+				}
+
+				return bytes, nil
```
Comment (richardpringle) [inline x/programs/runtime/import_state.go:96]: 👍

File: x/programs/cmd/simulator/build.rs (hunk 1 above)
Comment (iFrostizz) [reply x/programs/cmd/simulator/build.rs]: David took care of moving things back from the `v2` directory in this PR, so this addresses that!

File: x/programs/runtime/import_program.go [hunk 14]
```diff
@@ -17,10 +17,10 @@ const (
 )
 
 type callProgramInput struct {
-	ProgramID    ids.ID
-	FunctionName string
-	Params       []byte
-	Fuel         uint64
+	ProgramID ids.ID
+	Function  []byte
```
Comment (richardpringle) [inline x/programs/runtime/import_program.go:21]: Why was this changed?

File: x/programs/cmd/simulator/build.rs [hunk 15]
```diff
@@ -24,8 +24,8 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
 
     let current_dir = std::env::current_dir().unwrap();
 
-    let simulator_path = "bin/simulator";
-    let simulator_src = "simulator.go";
+    let simulator_path = "../bin/simulator";
+    let simulator_src = "../simulator.go";
```
[the remark this answers is not in the document]
Comment (iFrostizz) [reply x/programs/cmd/simulator/build.rs]: It's back, thank you ! (And I actually have no idea)

File: scripts/tests.unit.sh (hunk 3 above)
Comment (iFrostizz) [reply scripts/tests.unit.sh:20]: I totally forgot that `-v` was to invert the match, you're right, I will remove this

File: x/programs/runtime/import_program.go (hunk 14 above)
Comment (dboehm-avalabs) [reply x/programs/runtime/import_program.go:21]: Yeah, I think it makes more sense as a string.

File: x/programs/runtime/import_log.go
Comment (dboehm-avalabs) [inline x/programs/runtime/import_log.go:1]: Should undo this rename, turns out the debug tag did work.

File: x/programs/runtime/import_log_release.go
Comment (dboehm-avalabs) [inline x/programs/runtime/import_log_release.go:1]: Should undo this delete, the debug tag did work.

File: x/programs/cmd/simulator/cmd/config.go [hunk 16]
```diff
@@ -54,17 +53,17 @@ func newResponse(id int) *Response {
...
 	// The error message if available.
-	Error string `json:"error,omitempty" yaml:"error,omitempty"`
+	Error string `json:"error,omitempty"`
 }
 
 func (r *Response) Print() error {
 	jsonBytes, err := json.Marshal(r)
 	if err != nil {
-		return fmt.Errorf("failed to marshal response: %s", err)
+		return fmt.Errorf("failed to marshal response: %w", err)
```
[the remark this answers is not in the document]
Comment (iFrostizz) [reply x/programs/cmd/simulator/cmd/config.go:66]: It's adding new capabilities to the returned error so it can be manipulated by the callers, it seems that there is just no downsides: https://go.dev/wiki/ErrorValueFAQ

File: x/programs/cmd/simulator/cmd/program.go [hunk 17]
```diff
@@ -148,3 +150,35 @@ func programExecuteFunc(
...
+		case uint32:
+			bs := make([]byte, 4)
+			binary.LittleEndian.PutUint32(bs, v)
+			bytes = append(bytes, bs...)
+		default:
+			return nil, errors.New("unsupported data type")
+		}
+	}
+	return bytes, nil
+}
```
[the remark this answers is not in the document]
Comment (iFrostizz) [reply x/programs/cmd/simulator/cmd/program.go:177]: I'm happy to keep the scope as small as possible (even though the lints could have been done in a separate PR) and do this in another PR.

File: x/programs/cmd/simulator/vm/genesis/rules.go [hunk 18]
```diff
@@ -51,6 +52,16 @@ func (r *Rules) GetWindowTargetUnits() fees.Dimensions {
...
 
+func (*Rules) GetMaxActionsPerTx() uint8 {
+	// TODO
+	return 0
+}
+
+func (*Rules) GetMaxOutputsPerAction() uint8 {
+	// TODO
+	return 0
+}
```
[the remark this answers is not in the document]
Comment (iFrostizz) [reply x/programs/cmd/simulator/vm/genesis/rules.go:63]: Indeed, a panic is probably the best option here, thanks!

File: x/programs/runtime/import_program.go [hunk 19]
```diff
@@ -17,10 +17,10 @@ const (
 )
 
 type callProgramInput struct {
-	ProgramID    ids.ID
-	FunctionName string
-	Params       []byte
-	Fuel         uint64
+	ProgramID []byte
```
[the remark this answers is not in the document]
Comment (iFrostizz) [reply x/programs/runtime/import_program.go]: I think that it wasn't really intended, rolling back

File: x/programs/rust/wasmlanche-sdk/src/program.rs [hunk 20]
```diff
@@ -84,14 +84,14 @@ impl<K: Key> Program<K> {
     }
 }
 
-struct CallProgramArgs<'a, K> {
-    target_id: &'a Program<K>,
+struct CallProgramArgs<'a> {
+    target_id: &'a [u8],
```
[the remark this answers is not in the document]
Comment (iFrostizz) [reply x/programs/rust/wasmlanche-sdk/src/program.rs]: Rolling back this one too, I didn't noticed that the serialization was handling that

File: x/programs/runtime/import_program.go (hunk 14 above)
Comment (iFrostizz) [reply x/programs/runtime/import_program.go:21]: This was unintended, Rolling back!

File: x/programs/runtime/import_log.go
Comment (iFrostizz) [reply x/programs/runtime/import_log.go:1]: Done!

File: x/programs/runtime/import_log_release.go
Comment (iFrostizz) [reply x/programs/runtime/import_log_release.go:1]: Done!

File: x/programs/cmd/simulator/cmd/plan.go [hunk 21]
```diff
@@ -58,7 +56,7 @@ func (c *runCmd) New(parser *argparse.Parser, programIDStrMap map[string]string,
 	c.reader = reader
 }
 
-func (c *runCmd) Run(ctx context.Context, log logging.Logger, db *state.SimpleMutable, args []string) (*Response, error) {
+func (c *runCmd) Run(ctx context.Context, log logging.Logger, db *state.SimpleMutable, _ []string) (*Response, error) {
```
[the remark this answers is not in the document]
Comment (iFrostizz) [reply x/programs/cmd/simulator/cmd/plan.go:59]: The CI was yelling at me for those lints, thanks a lot to David for fixing them, it was probably ignored in the past
Will do in a separate PR next time!

File: x/programs/cmd/simulator/cmd/program.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (richardpringle) [reply x/programs/cmd/simulator/cmd/program.go:177]: https://github.com/ava-labs/hypersdk/issues/927
Thanks for creating the issue!

File: x/programs/cmd/simulator/vm/genesis/rules.go [hunk 22]
```diff
@@ -51,6 +52,14 @@ func (r *Rules) GetWindowTargetUnits() fees.Dimensions {
 	return r.g.WindowTargetUnits
 }
 
+func (*Rules) GetMaxActionsPerTx() uint8 {
+	panic("unimplemented")
+}
+
+func (*Rules) GetMaxOutputsPerAction() uint8 {
+	panic("unimplemented")
+}
```
Comment (richardpringle) [inline x/programs/cmd/simulator/vm/genesis/rules.go:61]: 🚀

**Review by richardpringle — DISMISSED** [review]

**Review by richardpringle — APPROVED** [review]
